src/game/client/water/w3dwatertracks.cpp

```diff
@@ -13,15 +13,324 @@
+                0, vertex_buffer->FVF_Info().Get_FVF_Size() * m_y * m_x, (BYTE **)&verts, D3DLOCK_DISCARD)) {
+            return batch_start;
+        }
+
+        batch_start = 0;
+    }
+
+    float alpha;
+    float f3 = 1.0f;
+
+    if (m_type == WAVE_TYPE_UNK) {
+        v2 = m_startPos;
+        v2 -= (m_perpDir * m_length) * 0.5f;
+        v1 = v2 - ((oo_wave_dir_len * m_width) * m_waveDir);
+        alpha = 0.0f;
+
+        if (m_elapsedMs >= m_totalMs) {
+            m_elapsedMs = 0;
+        }
+
+        if (m_timingUnk1 + m_timingUnk2 - 1000.0f + (float)m_fadeMs < (float)m_elapsedMs) {
+            alpha = (float)m_elapsedMs - (m_timingUnk1 + m_timingUnk2 - 1000.0f + (float)m_fadeMs);
+            alpha = alpha / m_timingUnk3;
+            alpha = 1.0f - alpha;
+
+            if (alpha < 0.0f) {
+                alpha = 0.0f;
+            }
+        } else if (m_timingUnk1 + m_timingUnk2 - 1000.0f < (float)m_elapsedMs) {
+            alpha = (float)m_elapsedMs - (m_timingUnk1 + m_timingUnk2 - 1000.0f);
+            alpha = alpha / (float)m_fadeMs;
+
+            if (alpha > 1.0f) {
+                alpha = 1.0f;
+            }
+        }
+    } else if ((float)m_elapsedMs < m_timingUnk1) {
+        alpha = (float)m_elapsedMs / m_timingUnk1;
+        f3 = alpha;
+        f3 = ((m_length - m_widthFraction) * f3 + m_widthFraction) / m_length;
+        v2 = m_startPos + (((float)m_elapsedMs * m_velocity * oo_wave_dir_len) * m_waveDir);
+        v2 -= ((m_perpDir * m_length) * 0.5f) * f3;
+        v1 = v2 - ((oo_wave_dir_len * m_heightFraction) * m_waveDir);
+    } else if (m_elapsedMs < m_totalMs) {
+        alpha = 1.0f;
+        f3 = 1.0f;
+        v2 = m_startPos + ((m_velocity * m_timingUnk1 * oo_wave_dir_len) * m_waveDir);
+        v1 = v2;
+        float f4 = (float)m_elapsedMs - m_timingUnk1;
+        v2 += ((f4 * m_velocity + 0.5f * m_velocityUnk * f4 * f4) * oo_wave_dir_len) * m_waveDir;
+        v2 -= ((m_perpDir * m_length) * 0.5f) * f3;
+        float f5 = (float)m_elapsedMs - m_timingUnk1 - m_timingUnk2;
+
+        if (f5 < 0.0f) {
+            f5 = 0.0f;
+        }
+
+        alpha = f5 / (float)m_fadeMs;
+
+        if (alpha > 1.0f) {
+            alpha = 1.0f;
+        }
+
+        alpha = 1.0f - alpha;
+        v1 -= (oo_wave_dir_len * m_heightFraction) * m_waveDir;
+
+        if (m_timingUnk1 + m_timingUnk2 + m_timeToCompress < (float)m_elapsedMs) {
+            v1 +=
+                (0.5f * m_scaleUnk * (m_timingUnk2 + m_timeToCompress) * (m_timingUnk2 + m_timeToCompress) * oo_wave_dir_len)
+                * m_waveDir;
+            float f6 = (float)m_elapsedMs - (m_timingUnk1 + m_timingUnk2 + m_timeToCompress);
+            v1 += (0.5f * m_velocityUnk * f6 * f6 * oo_wave_dir_len) * m_waveDir;
+        } else {
+            v1 += (0.5f * m_scaleUnk * f4 * f4 * oo_wave_dir_len) * m_waveDir;
+        }
+
+        v1 -= ((m_perpDir * m_length) * 0.5f) * f3;
+    } else {
+        m_elapsedMs = 0;

```

Comment:
This is just a repeat of (float)m_elapsedMs < m_timingUnk1 except it resets the elapsedMs first. I.e. this is restarting the loop.